### PR TITLE
fix(runtime): CPU thrashing when recovering from EAGAIN

### PR DESCRIPTION
**Pull request by RomainMuller**
The `node` runtime does not allow synchronous interactions with STDIN,
which is a problem because the jsii kernel **MUST ABSOLUTELY** have a
synchronous communication channel with the host process (otherwise, all
native overrides would result in an asynchronous **JavaScript** proxy,
which would break all **JavaScript** usage of those - since the original
code does not expect to have to deal with promises).

Unfortunately, we are not able to provision FIFOs (using `mkfifo`)
because those thiings do not exist on Windows. The closest thing to
FIFOs on Windows is IPC Channels, however `node` offers no synchronous
access to those either.

In order to solve this problem, we are leveraging `worker_threads` (a
feature that is available since around `node 9`, however gated by the
`--experimental-worker` feature flag in `node < 12` -- thankfully the
API has not really changed before it was stabilized, and we are in
control of how the `node` process is started).

The main thread (which the applications tarts on) create a `Worker`
thread and provides it with a shared memory buffer and the non-blocking
file descriptor (typically, of `process.stdin`, aka: `0`). The `Worker`
thread and main thread coordinate using the `Atomics` API, so that the
main thread is effectively blocked until the `Worker` thread has made
some data available to it (the `Worker` thread is then blocked until the
main thread has consumed all of that data).

Note that the `Atomics.wait` call suspends the whole agent that makes it
until the condition breaks or time-out is reached. This means absolutely
nothing else will happen on that thread during this sleep, including
processing other promises: from the moment code calls into the
`SyncReader.readSync` method, and until it has returned, no other
**JavaScript** code will be executed on the calling thread.

Fixes aws/aws-cdk#12069

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

**Comment by RomainMuller** [comment]
Note - we are planning a face-to-face review of this code with @NetaNir. Please do not get that PR merged before we had a chance to get there.

**Comment by RomainMuller** [comment]
Okay so a .NET test suite appears to hang for some reason. Troubleshooting this is going to take more time than I have today.

**Review by MrArnoldPalmer — COMMENTED** [review]
Just some general questions for this first roung.

Currently we support node [^10.3.0](https://aws.github.io/jsii/dev-guide/) though this would make the minimum 10.5.0. Is that an acceptable breakage for this or do we need a MV?

It looks like we are changing how we embed the JS into various runtimes. In some cases it seems to make the build process more windows friendly. Is that the only reason for these?

Additionally, do we know what the causes of the dotnet tests hanging were? It looks like we explicitly handle some new error cases (also in go). Was these errors being ignored the issue?

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -8,7 +8,8 @@ on:
     branches: [main, release]
 
 env:
-  DOTNET_NOLOGO: true
+  DOTNET_NOLOGO: true # Suppress logo & telemetry message from first run
+  DOTNET_CLI_TELEMETRY_OPTOUT: true # Disable dotnet CLI telemetry
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply .github/workflows/main.yml:12]: Parts of the errands... At this stage I don't have a clue anymore, so I'm willing to give ANYTHING a try.

File: packages/@jsii/dotnet-runtime/src/Amazon.JSII.Analyzers.UnitTests/JsiiOptionalAnalyzerTests.cs [hunk 2]
```diff
@@ -10,15 +11,15 @@ namespace Amazon.JSII.Analyzers.UnitTests
     public class JsiiOptionalAnalyzerTests : DiagnosticVerifier
     {
         [Fact]
-        public void TestGivenSomeEmptyCodeThenRoslynDoesNotComplain()
+        public async Task TestGivenSomeEmptyCodeThenRoslynDoesNotComplain()
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/@jsii/dotnet-runtime/src/Amazon.JSII.Analyzers.UnitTests/JsiiOptionalAnalyzerTests.cs:14]: This was supposed to possibly be what made the test suite hang... but no 😭

**Comment by RomainMuller** [comment]
> @MrArnoldPalmer:
>
> Currently we support node ^10.3.0 though this would make the minimum 10.5.0. Is that an acceptable breakage for this or do we need a MV?

I don't suppose many folks are still on `<10.5.0`, however this is a real question. It would be possible to alter this change in such a way that we use this new code _only_ if `worker_threads` is available (or enable-able), and fall-back to the hot-polling approach that this was set to replace....

This would allow us to keep the same compatibility at the expense of some mild additional complexity... I'm open to that, in fact... but is it worth it?

**Comment by RomainMuller** [comment]
Please leave the `pr/do-not-merge` tag as I want to explore a possibly widely less heavy-handed approach before we merge this monster.

**Comment by RomainMuller** [comment]
This PR is likely going to be superseded by #2411, which appears to be a superior solution (more reliable, less complex, and with more beneficial side-effects).

**Comment by RomainMuller** [comment]
I'm closing this one off, because #2411 is ready and I think it's much better a solution for this problem.

**Pull request closed without merging**